File: fast-chat/go/internal/chat/state.go
package chat

import (
	"log"
	"sync"
	"sync/atomic"
	"time"

	"github.com/google/uuid"
)

const (
	MaxMessages = 1000 // Circular buffer size
)

// ChatState holds the global chat state
type ChatState struct {
	messages     *RingBuffer // Lock-free ring buffer for messages
	connections  map[uuid.UUID]*Connection
	nicknamePool *NicknamePool
	logger       *MessageLogger
	nextID       atomic.Int64
	connsMu      sync.RWMutex // Protects connections map
}

// NewChatState creates a new chat state
func NewChatState(logFile string, flushInterval time.Duration) (*ChatState, error) {
	logger, err := NewMessageLogger(logFile, flushInterval)
	if err != nil {
		return nil, err
	}

	state := &ChatState{
		messages:     NewRingBuffer(MaxMessages),
		connections:  make(map[uuid.UUID]*Connection),
		nicknamePool: NewNicknamePool(),
		logger:       logger,
	}

	// Load messages from log file
	loadedMessages, err := LoadMessages(logFile)
	if err != nil {
		logger.Close()
		return nil, err
	}

	// Log the number of messages loaded
	totalLoaded := len(loadedMessages)
	if totalLoaded == 0 {
		log.Println("No previous messages found, starting fresh")
	} else {
		log.Printf("Loaded %d messages from log file", totalLoaded)
	}

	// Populate ring buffer with last MaxMessages
	if len(loadedMessages) > 0 {
		start := 0
		if len(loadedMessages) > MaxMessages {
			start = len(loadedMessages) - MaxMessages
			log.Printf("Keeping last %d messages in memory (discarding %d older messages)", MaxMessages, totalLoaded-MaxMessages)
		}

		// Push messages into ring buffer
		for _, msg := range loadedMessages[start:] {
			state.messages.Push(msg)
		}

		// Update nextID to continue from last message
		lastMsg := loadedMessages[len(loadedMessages)-1]
		state.nextID.Store(lastMsg.ID)
		log.Printf("Resuming message IDs from %d", lastMsg.ID)
	}

	return state, nil
}

// Close gracefully shuts down the chat state
func (s *ChatState) Close() error {
	if s.logger != nil {
		return s.logger.Close()
	}
	return nil
}

// AddMessage adds a message to the ring buffer (lock-free!)
func (s *ChatState) AddMessage(nickname, text string) *Message {
	id := s.nextID.Add(1)
	msg := NewMessage(id, nickname, text)

	// Add message to lock-free ring buffer (no lock needed!)
	s.messages.Push(msg)

	// Log message asynchronously
	if s.logger != nil {
		s.logger.Log(msg)
	}

	return msg
}

// GetLastN returns the last N messages (for initial page load)
// Lock-free read from ring buffer
func (s *ChatState) GetLastN(n int) []*Message {
	return s.messages.GetLast(n)
}

// GetHistory returns messages for history requests
// skip = how many to skip from the end, take = how many to return
// Lock-free read from ring buffer
func (s *ChatState) GetHistory(skip, take int) []*Message {
	return s.messages.GetHistory(skip, take)
}

// AddConnection adds a new connection
func (s *ChatState) AddConnection(conn *Connection) {
	s.connsMu.Lock()
	defer s.connsMu.Unlock()
	s.connections[conn.ID] = conn
}

// RemoveConnection removes a connection and releases its nickname
// Returns the nickname if it was removed, empty string if already removed
func (s *ChatState) RemoveConnection(id uuid.UUID) string {
	s.connsMu.Lock()
	defer s.connsMu.Unlock()

	conn, exists := s.connections[id]
	if !exists {
		return "" // Already removed
	}

	// Check if already closed (race condition protection)
	if conn.IsClosed() {
		delete(s.connections, id)
		return "" // Already handled
	}

	// Get nickname before closing
	nickname := conn.GetNickname()

	// Close and cleanup
	conn.Close()
	delete(s.connections, id)

	// Release nickname back to pool
	if nickname != "" {
		s.nicknamePool.Release(nickname)
	}

	return nickname
}

// GetConnection gets a connection by ID
func (s *ChatState) GetConnection(id uuid.UUID) (*Connection, bool) {
	s.connsMu.RLock()
	defer s.connsMu.RUnlock()
	conn, exists := s.connections[id]
	return conn, exists
}

// ConnectionCount returns the number of active connections
func (s *ChatState) ConnectionCount() int {
	s.connsMu.RLock()
	defer s.connsMu.RUnlock()
	return len(s.connections)
}

// Broadcast sends a message to all active connections
func (s *ChatState) Broadcast(msg string) {
	// Minimize lock hold time by copying connection pointers
	s.connsMu.RLock()
	conns := make([]*Connection, 0, len(s.connections))
	for _, conn := range s.connections {
		// Only include connections that aren't closed
		if !conn.IsClosed() {
			conns = append(conns, conn)
		}
	}
	s.connsMu.RUnlock()

	// Send to all active connections without holding the lock
	for _, conn := range conns {
		conn.Send(msg)
	}
}

// BroadcastExcept sends a message to all active connections except one
func (s *ChatState) BroadcastExcept(msg string, exceptID uuid.UUID) {
	// Minimize lock hold time by copying connection pointers
	s.connsMu.RLock()
	conns := make([]*Connection, 0, len(s.connections))
	for id, conn := range s.connections {
		// Only include connections that aren't closed and aren't the exception
		if id != exceptID && !conn.IsClosed() {
			conns = append(conns, conn)
		}
	}
	s.connsMu.RUnlock()

	// Send to all active connections without holding the lock
	for _, conn := range conns {
		conn.Send(msg)
	}
}

// AllocateNickname allocates a nickname from the pool
func (s *ChatState) AllocateNickname() string {
	return s.nicknamePool.Allocate()
}

// NicknamePoolStats returns stats about the nickname pool
func (s *ChatState) NicknamePoolStats() (available, used int) {
	return s.nicknamePool.Available(), s.nicknamePool.Used()
}

File: fast-chat/rust/src/chat/state.rs
use super::connection::Connection;
use super::logger::{archive_log_file, load_messages, MessageLogger};
use super::message::Message;
use super::nicknames::NicknamePool;
use super::ringbuffer::RingBuffer;
use anyhow::Result;
use parking_lot::RwLock;
use std::collections::HashMap;
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::Arc;
use std::time::Duration;
use uuid::Uuid;

pub struct ChatState {
    messages: Arc<RwLock<RingBuffer>>,
    connections: Arc<RwLock<HashMap<Uuid, Connection>>>,
    nickname_pool: Arc<RwLock<NicknamePool>>,
    logger: Arc<MessageLogger>,
    next_id: AtomicI64,
}

impl ChatState {
    pub async fn new(log_file: &str, max_messages: usize, log_buffer_size: usize, flush_interval: Duration) -> Result<Self> {
        let loaded_messages = load_messages(log_file).await?;

        if let Err(e) = archive_log_file(log_file).await {
            tracing::warn!("Failed to archive log file: {}", e);
        }

        let logger = MessageLogger::new(log_file, log_buffer_size, flush_interval).await?;

        let total_loaded = loaded_messages.len();
        if total_loaded == 0 {
            tracing::info!("No previous messages found, starting fresh");
        } else {
            tracing::info!("Loaded {} messages from log file", total_loaded);
        }

        let mut ring_buffer = RingBuffer::new(max_messages);
        let mut next_id = 0i64;

        if !loaded_messages.is_empty() {
            let start = if total_loaded > max_messages {
                tracing::info!(
                    "Keeping last {} messages in memory (discarding {} older messages)",
                    max_messages,
                    total_loaded - max_messages
                );
                total_loaded - max_messages
            } else {
                0
            };

            for msg in &loaded_messages[start..] {
                ring_buffer.push(msg.clone());
            }

            next_id = loaded_messages.last().unwrap().id;
            tracing::info!("Resuming message IDs from {}", next_id);
        }

        Ok(Self {
            messages: Arc::new(RwLock::new(ring_buffer)),
            connections: Arc::new(RwLock::new(HashMap::new())),
            nickname_pool: Arc::new(RwLock::new(NicknamePool::new())),
            logger: Arc::new(logger),
            next_id: AtomicI64::new(next_id),
        })
    }

    pub fn add_message(&self, nickname: String, text: String) -> Message {
        let id = self.next_id.fetch_add(1, Ordering::SeqCst) + 1;
        let msg = Message::new(id, nickname, text);

        self.messages.write().push(msg.clone());

        self.logger.log(msg.clone());

        msg
    }

    pub fn get_last_n(&self, n: usize) -> Vec<Arc<Message>> {
        self.messages.read().get_last(n)
    }

    pub fn get_history(&self, skip: usize, take: usize) -> Vec<Arc<Message>> {
        self.messages.read().get_history(skip, take)
    }

    pub fn add_connection(&self, conn: Connection) {
        self.connections.write().insert(conn.id, conn);
    }

    pub fn remove_connection(&self, id: Uuid) -> Option<String> {
        let mut connections = self.connections.write();
        
        if let Some(conn) = connections.get(&id) {
            if conn.is_closed() {
                connections.remove(&id);
                return None;
            }

            let nickname = conn.get_nickname();
            conn.close();
            connections.remove(&id);

            if !nickname.is_empty() {
                self.nickname_pool.write().release(nickname.clone());
                return Some(nickname);
            }
        }

        None
    }

    #[allow(dead_code)]
    pub fn get_connection(&self, id: Uuid) -> Option<Connection> {
        self.connections.read().get(&id).cloned()
    }

    pub fn connection_count(&self) -> usize {
        self.connections.read().len()
    }

    pub fn broadcast(&self, msg: String) {
        let msg = Arc::new(msg);
        let connections = self.connections.read();
        
        for conn in connections.values() {
            if !conn.is_closed() {
                conn.send(Arc::clone(&msg));
            }
        }
    }

    pub fn broadcast_except(&self, msg: String, except_id: Uuid) {
        let msg = Arc::new(msg);
        let connections = self.connections.read();
        
        for (id, conn) in connections.iter() {
            if *id != except_id && !conn.is_closed() {
                conn.send(Arc::clone(&msg));
            }
        }
    }

    pub fn allocate_nickname(&self) -> String {
        self.nickname_pool.write().allocate()
    }

    #[allow(dead_code)]
    pub fn nickname_pool_stats(&self) -> (usize, usize) {
        let pool = self.nickname_pool.read();
        (pool.available(), pool.used())
    }
}
